### PR TITLE
API: Default to hidden visibility for API tables

### PR DESCRIPTION
**Pull request by seberg**
This adds a new ``NPY_API_SYMBOL_ATTRIBUTE`` but defaults to using hidden visibility (always the case on windows!).

That actually makes the situation "worse" for ``eigenpy`` in some degree, since it forces them to adapt, but it also allows them to decide to just roll with it (by actually also exporting it on windows which).

Since it aligns windows and linux it seems like a good idea?  OTOH, who knows how many projects get away with just not caring about windows...

I tried to reorganize the docs a bit on how to import the array API...

---

~**Should be ready, but for discussion: we have to make a decision that this is better and ideally improves the ``eigenpy`` situation (if just be clarifying it).  EDIT: See gh-26098 and gh-26091.**~

### CODE REVIEW

**Comment by charris** [comment]
This should be backported?

**Comment by seberg** [comment]
Maybe?  It only really matters if we think it helps clarify the situation for 2.0 and I am not certain.
I.e. it forces/nudges users to not share the API table between shared libraries.  Something that is shaky on windows and could be an anti-pattern if the shared libraries are not all part of one project.

Maybe @mattip has an opinion on this?

Otherwise, I think this may be nice but if it causes breakage it is probably niche enough that we can get away with it in a 2.1 release.

**Comment by charris** [comment]
Sounds like it needs more experimentation, so 2.1 is probably the better option.

**Comment by rgommers** [comment]
This seems like a good idea to me. It prevents potential problems I think, and having Windows and other OSes behave consistently here has to be better.

Regarding symbol visibility, I'll note that:
- Meson turns on `-fvisibility=hidden` (`inlineshidden` for C++) by default: https://mesonbuild.com/Python-module.html#methods. Extension modules should only need the `PyInit_xxx` symbol. 
- SciPy uses a linker script in addition to really hide everything to the extent possible (originally implemented by Pauli to deal with symbol clashes in some corner cases): https://github.com/scipy/scipy/blob/7b3831a394d012176f95596df971b8bb7248d6c0/meson.build#L119-L131

**Comment by seberg** [comment]
Ah nice point, maybe another reason that adding it explicitly may be good to make downstream setups not accidentally rely on somtehing that is brittle (by build setup/system rules).
(I.e. if we create pain now, at least we are likely to safe pain elsewhere because eventually someone will run into it.)

> Meson turns on -fvisibility=hidden (inlineshidden for C++) 

I guess we don't actually use it?  Would be a nice cleanup (not *now*!) to remove all the `NPY_NO_EXPORT` and only mark the functions that we do want to export.

**Comment by seberg** [comment]
We discussed it and landed on trying this for 2.1 (which means that this should be fine but the release notes/migration changes need to be moved around a bit).

**Comment by rgommers** [comment]
> I guess we don't actually use it? Would be a nice cleanup (not _now_!) to remove all the `NPY_NO_EXPORT` and only mark the functions that we do want to export.

Cleanup would be nice indeed. We do use it, it's automatic. Try this (after you've built with `spin`):
```bash
$ meson introspect build --targets -i > targets.json
# now open targets.json and search for `hidden`
```

**Comment by seberg** [comment]
OK, finally moved the notes in the last commit.  This should be squash merged to avoid editing the release notes back and forth.

**Comment by seberg** [comment]
Rebased (hopefully correctly), it would be good to check the docs reorg (the technical change is pretty small).

File: doc/source/reference/c-api/array.rst [hunk 1]
```diff
@@ -3815,13 +3815,119 @@ Other conversions
...
+* If ``NO_IMPORT_ARRAY`` is defined, the table is declared as
+  ``extern void **``, meaning that it must be linked to a file which does not
+  use ``NO_IMPORT_ARRAY``.
+
+The ``PY_ARRAY_UNIQUE_SYMBOL`` mechanism additionally mangles the names to
+avoid conflicts.
+
+
+.. versionchanged::
+    NumPy 2.0 exports the headers to avoid sharing the table outside of a
```
Comment (ngoldbaum) [inline doc/source/reference/c-api/array.rst:3928]: Now that this PR is targeting numpy 2.1 it's a bit of a shame that these docs won't be available for the 2.0 release. Maybe consider splitting off into a separate PR if that's not obnoxious or manually backporting after this is merged?

Also now that this is targeting 2.1, I think the versionchanged here is 2.1. The other versionadded above is 2.0.

File: doc/source/reference/c-api/array.rst [hunk 2]
```diff
@@ -3845,26 +3951,41 @@ the C-API is needed then some additional steps must be taken.
...
+    ``PyArray_ImportNumPyAPI()`` is preferable for custom error checking.
+    You may also see use of ``_import_array()`` (a function, not
+    a macro, but you may want to raise a better error if it fails) and
+    the variations ``import_array1(ret)`` which customizes the return value.
 
 .. c:macro:: PY_ARRAY_UNIQUE_SYMBOL
 
+.. c:macro:: NPY_API_SYMBOL_ATTRIBUTE
+
+    .. versionadded:: 2.0
```
Comment (ngoldbaum) [inline doc/source/reference/c-api/array.rst:3964]: 2.1 here as well

**Review by ngoldbaum — COMMENTED** [review]
The new docs look good! I wish this was available when I was figuring out how the `import_array()` mechanism is supposed to work in late 2022.

**Comment by seberg** [comment]
In principle we could backport just the docs or even more (even if it says 2.1, I guess).  I guess we could even backport the whole thing, so long we change the default to do nothing.

Or just backport and delete all the new things, shouldn't be hard after all.

**Comment by ngoldbaum** [comment]
Sorry for letting this one drop, let's bring this in.

**Pull request merged**